compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp

```diff
@@ -150,6 +150,155 @@ LogicalResult verifySPIRVMatmulPromoteVectorizePassPipeline(
+      !llvm::isPowerOf2_64(workgroupSize[1]) ||
+      !llvm::isPowerOf2_64(workgroupSize[2])) {
+    return op->emitOpError(
+        "expected each workgroup size dimension to be power of two");
+  }
+
+  // Verify each dimension of workgroup size should not exceed the corresponding
+  // limit of maxWorkGroupSize
+  if (workgroupSize[0] > maxWorkGroupSize[0] ||
+      workgroupSize[1] > maxWorkGroupSize[1] ||
+      workgroupSize[2] > maxWorkGroupSize[2]) {
+    return op->emitOpError("expected workgroup size dimensions not exceeding ")
+           << "[" << maxWorkGroupSize[0] << ", " << maxWorkGroupSize[1] << ", "
+           << maxWorkGroupSize[2] << "]";
+  }
+
+  // Verify the total workgroup size should not exceed maxThreads
+  int64_t totalWorkgroupSize =
+      workgroupSize[0] * workgroupSize[1] * workgroupSize[2];
+  if (totalWorkgroupSize > maxThreads) {
+    return op->emitOpError(
+               "expected total invocation count in workgroup to be <= ")
+           << maxThreads << ", got " << totalWorkgroupSize;
+  }
+
+  // Verify the total workgroup size should be multiple of subgroupSize
+  if (totalWorkgroupSize % subgroupSize != 0) {
+    return op->emitOpError("expected total workgroup size to be multiple of ")
+           << subgroupSize;
+  }
+
+  // Verify the total workgroup size should be equal or larger than 2 *
+  // subgroupSize
+  if (totalWorkgroupSize / subgroupSize < 2) {
+    return op->emitOpError("expected total workgroup size to be >= ")
+           << 2 * subgroupSize;
+  }
+
+  Type inputType = op->getOperand(0).getType();
+  ArrayRef<int64_t> lhsShape =
+      op->getOperand(0).getType().cast<ShapedType>().getShape();
+  ArrayRef<int64_t> rhsShape =
+      op->getOperand(1).getType().cast<ShapedType>().getShape();
+
+  SmallVector<int64_t> firstLevelTileSizes =
+      loweringConfig.getTileSizeVals(kWorkgroupTileLevel);
+  SmallVector<int64_t> secondLevelTileSizes =
+      loweringConfig.getTileSizeVals(kThreadTileLevel);
+  SmallVector<int64_t> thirdLevelTileSizes =
+      loweringConfig.getTileSizeVals(kReductionTileLevel);
+
+  if (loweringConfig.getTileSizes().size() != 4) {
+    return op->emitOpError("expected 4 levels of tiling sizes, got ")
+           << loweringConfig.getTileSizes().size();
+  }
+
+  // For BatchMatmul, the first dimension is the batch dimension.
+  // We don't check the batch.
+  if (isa<linalg::BatchMatmulOp>(op)) {
+    lhsShape = lhsShape.drop_front(1);
+    rhsShape = rhsShape.drop_front(1);
+    firstLevelTileSizes.erase(firstLevelTileSizes.begin());
+    secondLevelTileSizes.erase(secondLevelTileSizes.begin());
+    thirdLevelTileSizes.erase(thirdLevelTileSizes.begin());
+  }
+
+  // Verify that subgroup tile sizes should be multiple of 16
+  if (secondLevelTileSizes[0] % 16 != 0 || secondLevelTileSizes[1] % 16 != 0 ||
+      thirdLevelTileSizes[2] % 16 != 0) {
+    return op->emitOpError("expected subgroup tile sizes to be multiple of 16");
+  }
+
```

Comment:
Define `const int64_t dimM = lhsShape[0], dimN = rhsShape[1], dimK = lhsShape[1];` to make it more readable?